### PR TITLE
Improve Batch Comparison Audit Report 

### PR DESCRIPTION
**Pull request by carolinemodic**
https://github.com/votingworks/arlo/issues/1909
https://github.com/votingworks/arlo/issues/1784

Adds the following items to the audit report: 

For the "rounds" section - for ballot comparision audits
- number of batches - distinct number of batches (non distinct if batches are sampled multiple times is already included as the sample size) 
- number of ballots - number of ballots in distinct batches (could be changed to non-distinct or that could be additionally included) 
- reported votes - aggregated reported votes we expect for all sampled batches (deduped) 

I did not add a totals row - felt overly complex with having to thing about how everything should sum for all ballot types and overall fairly low-pri

For the "batches" section
- number of  ballots 

Tested via inspecting snapshot updates and testing manually with the small election fixtures. i had a batch sampled twice in my test to make sure the deduping bits were correct.

### CODE REVIEW

File: server/tests/ballot_comparison/snapshots/snap_test_cvrs.py [hunk 1]
```diff
@@ -7,1990 +7,7 @@
...
-            "Choice 2-3": {"column": 4, "num_votes": 3},
-            "Write-In": {"column": 5, "num_votes": 1},
-        },
-        "total_ballots_cast": 14,
-        "votes_allowed": 1,
-    },
-}
-
-snapshots["test_hart_cvr_upload_with_scanned_ballot_information 3"] = [
+snapshots["test_hart_cvr_upload_with_duplicate_batch_names 1"] = [
```
Comment (carolinemodic) [inline server/tests/ballot_comparison/snapshots/snap_test_cvrs.py:10]: I don't really understand why anything in this file changed but I think the snapshots were just out of date ?

File: server/api/reports.py [hunk 2]
```diff
@@ -391,6 +391,11 @@ def round_rows(election: Election):
             ["Audited Votes: CVR", "Audited Votes: Non CVR"]
             if election.audit_type == AuditType.HYBRID
             else []
+        )
+        + (
+            ["# Batches", "# Ballots", "Reported Votes"]
```
Comment (jonahkagan) [inline server/api/reports.py:396]: I think the `#` is a tiny bit confusing. How about "Batches in Sample" and "Ballots in Sample" or "Batches Sampled" and "Ballots Sampled"

File: server/api/reports.py [hunk 3]
```diff
@@ -441,6 +446,36 @@ def round_rows(election: Election):
...
+                    if contest.id in batch.jurisdiction.batch_tallies[batch.name]
+                    else None
+                )
+                if reported_results_for_batch is not None:
+                    for choice in contest.choices:
+                        reported_results[choice.id] += reported_results_for_batch[
+                            choice.id
+                        ]
+
+            reported_results_by_name = reported_results and {
```
Comment (jonahkagan) [inline server/api/reports.py:475]: I think `reported_results` always exists, so this check is unnecessary

File: server/api/reports.py [hunk 4]
```diff
@@ -441,6 +446,36 @@ def round_rows(election: Election):
...
+                .all()
+            )
+            num_distinct_batches = len(distinct_batches)
+            num_distinct_ballots = sum(batch.num_ballots for batch in distinct_batches)
+
+            # get the reported votes aggregated across batches
+            reported_results: dict = {choice.id: 0 for choice in contest.choices}
+            for batch in distinct_batches:
+                reported_results_for_batch = (
+                    batch.jurisdiction.batch_tallies[batch.name][contest.id]
```
Comment (jonahkagan) [inline server/api/reports.py:465]: Accessing `batch.jurisdiction` will make a separate query for each batch's jurisdiction. Could be a bit expensive. Often I will profile API endpoints using Heroku/Sentry tools after we deploy them to identify bottlenecks. (Report generation is already slow and should probably be backgrounded)

We might be able to pull out just the reported vote count we need for each batch at the db level if we decide to optimize it.

File: server/api/reports.py [hunk 5]
```diff
@@ -441,6 +446,36 @@ def round_rows(election: Election):
...
+                    if contest.id in batch.jurisdiction.batch_tallies[batch.name]
+                    else None
+                )
```
Comment (jonahkagan) [inline server/api/reports.py:468]: Consider: `batch.jurisdiction.batch_tallies[batch.name].get(contest.id)`, which will return `None` if the key is missing from the dict

**Review by jonahkagan — APPROVED** [review]

File: server/api/reports.py (hunk 4 above)
Comment (carolinemodic) [reply server/api/reports.py:465]: Yeah I was wondering about perf. This data is already accessed/fetched to populate the "batches" section of the csv for batch-comparison audits (where I copied the logic from), but if thats already slow its still not great to add it again.

Comment (carolinemodic) [reply server/api/reports.py:465]: What data do you use to profile?

Comment (jonahkagan) [reply server/api/reports.py]: Oh interesting, if it's already loaded once then maybe we can just reuse it? I guess that might take some finagling

To profile, I use Sentry transaction tracing and Heroku database analytics

Comment (carolinemodic) [reply server/api/reports.py]: Do you just find some large actual audit that has run to test with though?

Comment (jonahkagan) [reply server/api/reports.py]: Both of those tools work by sampling queries from actual usage, so I usually just look at those periodically to see what's slow. It's less that I'll test a specific change that I just deployed.

Load testing specific changes I usually do in my dev env with some large audit test data. But then you don't have as useful profiling tools.

Sorry, I realize my initial comment was a bit misleading about how I use those tools.

Comment (carolinemodic) [reply server/api/reports.py]: Ah got it, I guess I wasn't sure if there was a pattern of testing on fake data at times like right now when audits aren't really actively running to help monitor usage

**Pull request merged**